### DISCUSSION TITLE
DailyPartitionsDefinition and DynamicPartitionsDefinition upstream-downstream relation

### DISCUSSION THREAD

**Discussion by ssillaots-boku** [dagster-io/dagster#19945] [Q&A]
Hey!
I want to create a pipeline where upstream consists of `DailyPartitionsDefinition` and `DynamicPartitionsDefinition` and downstream is `DynamicPartitionsDefinition`.

So far I've managed to create a pipeline where `MultiPartitionsDefinition` (Daily and Dynamic) downstream asset depends on 1 `MultiPartitionsDefinition` (Daily and Dynamic) assets and 1 `DailyPartitionsDefinition` asset.

Let's say downstream asset has week's worth of daily partitions and one dynamic partition. So 7 partitions. Now I'll create a new upstream partition called `another_one` for day `2023-02-20`. Since upstream and downstream DynamicPartitionsDefinition share the same partitions definition name then the downstream asset's "another_one" partition will be created automatically. Even though I only materialized that partition for `2023-02-20` then the downstream will have "availability" to materialize `another_one` for all days in that asset. So now I go from 7 partitions to 14 partitions. Not from 7 to 8. This will eventually create an insufferable amount of partitions.

Is there another way to approach this kind of workflow?
So far I haven't found a `PartitionMapping` which would work where upstream is `DailyPartitionsDefinition` and `DynamicPartitionsDefinition` and downstream is `DynamicPartitionsDefinition`.

```
@asset(partitions_def=HourlyPartitionsDefinition(datetime(2024, 2, 14)),
       auto_materialize_policy=AutoMaterializePolicy.eager(),
       group_name='dynamic')
def hourly_job(context: AssetExecutionContext):
    return


multi = MultiPartitionsDefinition(
    {
        "time": DailyPartitionsDefinition(start_date="2024-02-14"),
        "dynamic": DynamicPartitionsDefinition(name="d_1"),
    }
)


@asset(partitions_def=multi,
       io_manager_key=NO_OP_IO_MANAGER,
       group_name='dynamic')
def upstream_dynamic_asset_1(context: AssetExecutionContext):
    DagsterInstance.get().add_dynamic_partitions(partitions_def_name='down_def',
                                                 partition_keys=[context.partition_key])
    return



@asset(deps=[upstream_dynamic_asset_1,
             hourly_job],
       partitions_def=multi,
       auto_materialize_policy=AutoMaterializePolicy.eager(),
       io_manager_key=NO_OP_IO_MANAGER,
       group_name='dynamic')
def downstream_asset(context: AssetExecutionContext):
    return

```

**Comment by ssillaots-boku** [comment]
No ideas for this? :(

**Comment by OwenKephart** [comment]
Hi @ssillaots-boku -- what is the exact relationship that you want between these partitions? It's true that the MultiPartitionsDefinition takes the cross product of the partitions definitions it contains, but I'm not sure if I see a particular alternative here. Which combinations of dynamic partitions / time partitions would you consider to be "valid" here? Is it based on the creation date of the dynamic partition (i.e. they can only correspond to dates at or after the dynamic partition was created?). 

I'm not sure if this is relevant to your use case, but might make sense to just have a single dynamic dimension and just add some extra semantic information inside your partition keys (e.g. you could name your partitions like `another_one_2024-02-20` when you create them).

**Reply by ssillaots-boku** [reply]
Hi @OwenKephart 
My first idea would be that instead of n:n relation it would be 1:n. So you wouldn't have all days for all dynamic partitions but rather for every day you would have partitions that were materialized in that day. Visual representation:
```
2023-03-01 dynamic_1
           dynamic_2
2023-03-02 dynamic_2

right now it's
dynamic_1  2023-03-01
           2023-03-02
dynamic_2  2023-03-01
           2023-03-02            
```

> They can only correspond to dates at or after the dynamic partition was created?

This makes sense but it would still clutter up the "space" moving in some time. You would have a lot of unmaterialized partitions 
But also I'm not sure how to solve it. I'm just approaching it from my edge case's point of view.

> I'm not sure if this is relevant to your use case, but might make sense to just have a single dynamic dimension and just add some extra semantic information inside your partition keys (e.g. you could name your partitions like another_one_2024-02-20 when you create them).

That wouldn't help because I want to have a daily upstream and in this case AutoMaterializePolicy will look for a partition called `another_one_2024-02-20` which obviously doesn't exist.

**Reply by OwenKephart** [reply]
I see -- unfortunately I don't think there's a solution here that would fit those exact specifications. The best option here may be to go with the dynamic partition approach (`another_one_2024-02-20`) for the upstream asset, and then simply have the downstream asset have an AllPartitionMapping with respect to the daily-partitioned asset and have logic in the body of the downstream asset to actually only load a single day's partition based on the multi partition key.

**Reply by ssillaots-boku** [reply]
Right now I created a custom sensor for the downstream asset (which checks dynamic and daily upstream). It's a hackish solution but imo a bit better than other options.
Thank you for your thoughts!
